axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/SaleOrderInvoiceServiceImpl.java

```diff
@@ -953,6 +953,34 @@ public void displayErrorMessageBtnGenerateInvoice(SaleOrder saleOrder) throws Ax
     }
   }
 
+  @Override
+  public int getInvoicingState(SaleOrder saleOrder) {
+    int invoicingState = 0;
+    if (saleOrder.getAmountInvoiced().compareTo(BigDecimal.ZERO) > 0
+        && saleOrder.getAmountInvoiced().compareTo(saleOrder.getExTaxTotal()) < 0) {
+      invoicingState = 2;
+    }
+    if (saleOrder.getAmountInvoiced().compareTo(BigDecimal.ZERO) > 0
+            && (saleOrder.getAmountInvoiced().compareTo(saleOrder.getExTaxTotal()) == 0)
+        || saleOrder.getAmountInvoiced().compareTo(saleOrder.getExTaxTotal()) > 0) {
+      invoicingState = 3;
+    }
+    if (saleOrder.getAmountInvoiced().compareTo(BigDecimal.ZERO) == 0) {
+      if (atLeastOneInvoiceIsVentilated(saleOrder)
+          && saleOrder.getExTaxTotal().compareTo(BigDecimal.ZERO) == 0) {
+        invoicingState = 3;
+      } else {
+        invoicingState = 1;
+      }
+    }
+
+    return invoicingState;
+  }
+
+  protected boolean atLeastOneInvoiceIsVentilated(SaleOrder saleOrder) {
+    return getInvoices(saleOrder).stream().anyMatch(invoice -> invoice.getStatusSelect() == 3);
```

Comment:
Please change this to a `.count()` query and use static variable from InvoiceRepository instead of writing `3`